# Assignment 2
### Understanding Uncertainty
### Due 9/12

In [2]:
import pandas as pd

In [3]:
idf = pd.read_csv('C:/Users/school/understanding_uncertainty/data/iowa.csv') # Iowa Liquor Sales Data

In [4]:
idf.head(3)

Invoice/Item Number        Date  Store Number                   Store Name  \
0     INV-59108400026  06/06/2023          3723           J D SPIRITS LIQUOR   
1        S16879800057  01/16/2014          3926  LIQUOR DOWNTOWN / IOWA CITY   
2     INV-05301100019  06/05/2017          3829     GARY'S FOODS / MT VERNON   

  Zip Code          Category Name         Vendor Name  Item Number  \
0    51040  STRAIGHT RYE WHISKIES    INFINIUM SPIRITS        27102   
1    52240         VODKA 80 PROOF  HEAVEN HILL BRANDS        35416   
2    52314      CANADIAN WHISKIES     DIAGEO AMERICAS        11296   

         Item Description  Bottle Volume (ml)  State Bottle Cost  \
0       TEMPLETON RYE 4YR                 750              18.09   
1  BURNETT'S VODKA 80 PRF                 750               4.84   
2             CROWN ROYAL                 750              15.59   

   State Bottle Retail  Bottles Sold  Sale (Dollars)  
0                27.14             2           54.28  
1                 7.26            12           87.12  
2                23.39             6          135.66

1. 'State Bottle Retail' is the price at which the bottle is sold, while 'State Bottle Cost' is the price the state pays for a bottle.
- How do you define the most profitable item? Why? Explain your thinking.
- Use `.groupby` calculations to determine the most profitable liquors sold in Iowa.
- Make a plot of the top liquors.

Question 1 part 1:

THere are a few ways one could determine which liqour is the most profitable, depending on factors like each type of bottle sold or combinind ml of bottles of the same type and then calculating it that way, but in my case I am going to use bottles sold because by my own definition, even if two bottles have the same type of liquor, the different mls per bottle make them completely different items. 

So I would do (retail - cost) * bottles sold by item number to find the most profitable item. 

Question 1 part 2

In [9]:
#find actual profit per bottle sold

idf['profit_per_bottle'] = idf['State Bottle Retail'] - idf['State Bottle Cost']

In [10]:
#overall profit per item
idf['profit'] = idf['profit_per_bottle'] * idf['Bottles Sold']

In [14]:
idf['profit'].groupby([idf['Item Description'], idf['Bottle Volume (ml)'], idf['Item Number']]).sum().sort_values(ascending=False).head()

Item Description      Bottle Volume (ml)  Item Number
BLACK VELVET          1750                11788          190115.15
TITOS HANDMADE VODKA  1750                38178          151369.56
                      1000                38177          115513.14
CROWN ROYAL           750                 11296           80490.17
TITOS HANDMADE VODKA  750                 38176           79783.64
Name: profit, dtype: float64

most profitable item: 1750 ml black velvet, followed by 1750 ml titos handmade vodka, item 38177, 750 ml crown royale, and 750 ml tito's handmade vodka

2. This question is to further explore broadcasting in NumPy.

- In the first code chunk below, there are two vectors, $x$ and $y$. What happens when you try to subtract them?
- Instead, try `x.reshape(-1,1)-y.reshape(-1,1)`. What do you get?
- Try `x.reshape(1,-1)-y.reshape(1,-1)`. Compare the results to your previous answer.
- Write two double for-loops to replicate the results of the previous two answers. 
- Clearly explain how `x.reshape(1,-1)` and `x.reshape(-1,1)` change $x$

In [1]:
import numpy as np

x = np.array([1,3,5])
y = np.array([2,5])

3. The indicator function $\mathbb{I}\{ x \le z \}$ takes the value 0 if $x>z$ is false and 1 if $x \le z$ is true.

- For $y = 3$, plot the indicator function for any $z$
- Explain how $ 1- \mathbb{I}\{ x \le z \} = \mathbb{I}\{ x > z \}$. Drawing a plot might help, or looking at the definition that
$$
\mathbb{I}\{x \le z\} = \begin{cases}
0, &x > z \\
1, &x \le z
\end{cases}
$$
- The code below generates a random sample for different numbers of observations, $N = 3, 30, 300, 3000 $. Write a function to compute the ECDF curve for each sample, and plot your results.
- Describe in your own words what happens to the ECDF qualitatively as the sample size $N$ increases

In [4]:
import numpy as np
import seaborn as sns

rng = np.random.default_rng(100)

for N in (3,30,300,3000):
    X = rng.normal(0,1,N)
    # ECDF plot goes here:


4. A **quantile** generalizes the concept of the median so that instead of 50% of the mass below and 50% above, the $q$-quantile is the value for which $q$ proportion of the data are below that value and $(1-q)$ proportion of the data are above that value. This is a concept of **order**.

- Write a function that computes any quantile without using Numpy's `np.quantile`
- Write a function that computes the **Interquartile Range (IQR)** -- the .25 quantile and the .75 quantile, which brackets 50% of the data -- and the **whiskers**: $\pm 1.5 \times \text{IQR}$ from the edges of the IQR
- Write a function that computes a five-number summary and values for a boxplot: the minimum, the .25 quantile, the median, the .75 quantile, and the maximum, and the whiskers
- Anything outside the whiskers is an **outlier**; write a function that returns a Boolean vector indicating if the observations are outliers.
- For a numeric variable of your choice from one of the datasets, apply your functions from above, and compare your answers with `sns.boxplot` and `df.describe`. 

5. The kernel density estimator captures the concentration of the data around various points, and is defined as
$$
\hat{f}_{N,h}(z) = \frac{1}{N} \sum_{i=1}^N \frac{1}{h} k \left( \frac{x_i - z}{h}\right)
$$
where $h$ is the bandwidth and $k()$ is one of the kernel functions from the class materials (Gaussian, Epanechnikov, uniform, triangular).

- Write a function that takes the bandwidth $h$ as given, and computes and plots $\hat{f}_{N,h}(z)$, for the kernels listed in the slides
- Write a function that computes the plug-in bandwidth
- Use your functions to analyze some numeric variables from a dataset of your choice, using a variety of bandwidths
- Compare bandwidths that are greater or less than the plug-in value with the results for the plug-in value. When does the KDE appear to be overfitting? Underfitting?